clvm/operators.py

```diff
@@ -35,6 +49,97 @@
 }
 
 
+def args_len(op_name, args):
+    for arg in args.as_iter():
+        if arg.pair:
+            raise EvalError("%s requires int args" % op_name, arg)
+        yield len(arg.as_atom())
+
+
+# unknown ops are reserved if they start with 0xffff
+# otherwise, unknown ops are no-ops, but they have costs. The cost is computed
+# like this:
+
+# byte index (reverse):
+# n | .... | 1        | 0        |
+# --+- - --+------+---+----------+
+# n | .... |      |fun| ignored  |
+# --+- - --+------+---+----------+
+# multiplier      |
+#                 | 2 bits
+
+# 1 is always added to the multiplier before using it to multiply the cost, this
+# is since cost may not be 0.
+
+# fun is 2 bits and defines how cost is computed based on arguments:
+# 0: constant, cost is 1 * (multiplier + 1)
+# 1: computed like operator add, multiplied by (multiplier + 1)
+# 2: computed like operator mul, multiplied by (multiplier + 1)
+# 3: computed like operator concat, multiplied by (multiplier + 1)
+
+# this means that unknown ops where fun is 1, 2, or 3, may still be fatal errors
+# if the arguments passed are not atoms.
+
+def default_unknown_op(op: bytes, args: CLVMObject) -> Tuple[int, CLVMObject]:
+    # any opcode starting with ffff is reserved (i.e. fatal error)
+    if len(op) >= 2 and op[:2] == b"\xff\xff":
+        raise EvalError("reserved operator", args.to(op))
+
+    # all other unknown opcodes are no-ops
+    # the cost of the no-ops is determined by the opcode number, except the
+    # least significant byte. unknown opcodes of 1 byte have cost 1.
+    if len(op) <= 1:
+        return (1, SExp.null())
```

Comment:
We might want some variance in 1-byte opcodes, since they are the most useful to repurpose (since they are small in the stream, especially if < 128).

Reply:
yeah, I seem to recall someone suggesting a full byte, but I may be mistaken. Maybe it would be better to move the 2 bits of `cost_function` into the last byte, and just have 6 bits unused for purposes of computing costs.